src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4917,5 +4917,57 @@ public async Task<int> DisposeAsync()
             comp.VerifyDiagnostics();
             CompileAndVerify(comp, expectedOutput: "MoveNextAsync DisposeAsync Done");
         }
+
+        [Fact]
+        [WorkItem(30956, "https://github.com/dotnet/roslyn/issues/30956")]
+        public void GetAwaiterBoxingConversion()
+        {
+            var source =
+@"using System;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+interface I1 { }
+interface I2 { }
+struct StructAwaitable1 : I1 { }
+struct StructAwaitable2 : I2 { }
+
+class Enumerable
+{
+    public Enumerator GetAsyncEnumerator() => new Enumerator();
+    internal class Enumerator
+    {
+        public object Current => null;
+        public StructAwaitable1 MoveNextAsync() => new StructAwaitable1();
+        public StructAwaitable2 DisposeAsync() => new StructAwaitable2();
+    }
+}
+
+static class Extensions
+{
+    internal static TaskAwaiter<bool> GetAwaiter(this I1 x)
+    {
+        if (x == null) throw new ArgumentNullException(nameof(x));
```

Comment:
>if (x == null) throw new ArgumentNullException(nameof(x)); [](start = 8, length = 58)

Consider printing something so that we can verify that the methods are called.